File: sherpa-onnx/java-api/src/main/java/com/k2fsa/sherpa/onnx/OnlineSpeechDenoiser.java
// Copyright 2026 Xiaomi Corporation

package com.k2fsa.sherpa.onnx;

public class OnlineSpeechDenoiser {
    private long ptr = 0;

    public OnlineSpeechDenoiser(OnlineSpeechDenoiserConfig config) {
        LibraryLoader.maybeLoad();
        ptr = newFromFile(config);
        if (ptr == 0) {
            throw new IllegalArgumentException("Invalid OnlineSpeechDenoiserConfig: failed to create native OnlineSpeechDenoiser");
        }
    }

    public int getSampleRate() {
        return getSampleRate(ptr);
    }

    public int getFrameShiftInSamples() {
        return getFrameShiftInSamples(ptr);
    }

    public DenoisedAudio run(float[] samples, int sampleRate) {
        return run(ptr, samples, sampleRate);
    }

    public DenoisedAudio flush() {
        return flush(ptr);
    }

    public void reset() {
        reset(ptr);
    }

    @Override
    protected void finalize() throws Throwable {
        release();
    }

    public void release() {
        if (this.ptr == 0) {
            return;
        }
        delete(this.ptr);
        this.ptr = 0;
    }

    private native void delete(long ptr);

    private native int getSampleRate(long ptr);

    private native int getFrameShiftInSamples(long ptr);

    private native DenoisedAudio run(long ptr, float[] samples, int sampleRate);

    private native DenoisedAudio flush(long ptr);

    private native void reset(long ptr);

    private native long newFromFile(OnlineSpeechDenoiserConfig config);
}

File: sherpa-onnx/kotlin-api/OnlineSpeechDenoiser.kt
package com.k2fsa.sherpa.onnx

import android.content.res.AssetManager

data class OnlineSpeechDenoiserConfig(
    var model: OfflineSpeechDenoiserModelConfig = OfflineSpeechDenoiserModelConfig(),
)

class OnlineSpeechDenoiser(
    assetManager: AssetManager? = null,
    config: OnlineSpeechDenoiserConfig,
) {
    private var ptr: Long

    init {
        ptr = if (assetManager != null) {
            newFromAsset(assetManager, config)
        } else {
            newFromFile(config)
        }
    }

    protected fun finalize() {
        if (ptr != 0L) {
            delete(ptr)
            ptr = 0
        }
    }

    fun release() = finalize()

    fun run(samples: FloatArray, sampleRate: Int) = run(ptr, samples, sampleRate)

    fun flush() = flush(ptr)

    fun reset() = reset(ptr)

    val sampleRate
      get() = getSampleRate(ptr)

    val frameShiftInSamples
      get() = getFrameShiftInSamples(ptr)

    private external fun newFromAsset(
        assetManager: AssetManager,
        config: OnlineSpeechDenoiserConfig,
    ): Long

    private external fun newFromFile(
        config: OnlineSpeechDenoiserConfig,
    ): Long

    private external fun delete(ptr: Long)

    private external fun run(ptr: Long, samples: FloatArray, sampleRate: Int): DenoisedAudio

    private external fun flush(ptr: Long): DenoisedAudio

    private external fun reset(ptr: Long)

    private external fun getSampleRate(ptr: Long): Int

    private external fun getFrameShiftInSamples(ptr: Long): Int

    companion object {
        init {
            System.loadLibrary("sherpa-onnx-jni")
        }
    }
}
